extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts

```diff
@@ -0,0 +1,348 @@
+			let serverInfo = await sqlops.connection.getServerInfo(connectionProfile.id);
+			if (!serverInfo || !credentials || !serverInfo.options) {
+				return false;
+			}
+			let endpoints: IEndpoint[] = serverInfo.options[constants.clusterEndpointsProperty];
+			if (!endpoints) {
+				return false;
+			}
+			let index = endpoints.findIndex(ep => ep.serviceName === constants.hadoopKnoxEndpointName);
+			if (!index) {
+				return false;
+			}
+
+			let connInfo: sqlops.connection.Connection = {
+				options: {
+					'host': endpoints[index].ipAddress,
+					'groupId': connectionProfile.options.groupId,
+					'knoxport': endpoints[index].port,
+					'user': 'root', //connectionProfile.options.userName cluster setup has to have the same user for master and big data cluster
+					'password': credentials.password,
+				},
+				providerName: constants.mssqlClusterProviderName,
+				connectionId: UUID.generateUuid()
+			};
+
+			let connection = new Connection(connInfo);
+			connection.saveUriWithPrefix(constants.objectExplorerPrefix);
+			let session = new Session(connection, sessionInfo.sessionId);
+			session.root = new RootNode(session, new TreeDataContext(this.appContext.extensionContext, this), sessionInfo.rootNode.nodePath);
+			this.sessionMap.set(sessionInfo.sessionId, session);
+			return true;
+		}
+	}
+
+	expandNode(nodeInfo: sqlops.ExpandNodeInfo, isRefresh: boolean = false): Thenable<boolean> {
+		return new Promise((resolve, reject) => {
+			if (!nodeInfo) {
+				reject('expandNode requires a nodeInfo object to be passed');
+			} else {
+				resolve(this.doExpandNode(nodeInfo, isRefresh));
+			}
+		});
+	}
+
+	private async doExpandNode(nodeInfo: sqlops.ExpandNodeInfo, isRefresh: boolean = false): Promise<boolean> {
+		let session = this.sessionMap.get(nodeInfo.sessionId);
+		let response = {
+			sessionId: nodeInfo.sessionId,
+			nodePath: nodeInfo.nodePath,
+			errorMessage: undefined,
+			nodes: []
+		};
+
+		if (!session) {
+			// This is not an error case. Just fire reponse with empty nodes for example: request from standalone SQL instance
+			this.expandCompleteEmitter.fire(response);
+			return false;
+		} else {
+			setTimeout(() => {
+
+				// Running after promise resolution as we need the Ops Studio-side map to have been updated
+				// Intentionally not awaiting or catching errors.
+				// Any failure in startExpansion should be emitted in the expand complete result
+				// We want this to be async and ideally return true before it completes
+				this.startExpansion(session, nodeInfo, isRefresh);
+			}, 10);
+		}
+		return true;
+	}
+
+	private async startExpansion(session: Session, nodeInfo: sqlops.ExpandNodeInfo, isRefresh: boolean = false): Promise<void> {
+		let expandResult: sqlops.ObjectExplorerExpandInfo = {
+			sessionId: session.uri,
+			nodePath: nodeInfo.nodePath,
+			errorMessage: undefined,
+			nodes: []
+		};
+		try {
+			let node = await session.root.findNodeByPath(nodeInfo.nodePath, true);
```

Comment:
This is so much better, thanks for taking the time to get this design right. Really appreciate how clean the new implementation is.